File: Stack/394. Decode String/Solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        # N : len(s)
        # Time Complexity; O(N)
        # Space Complexity; O(N)
        n = len(s)
        decodeStack = []
        i = 0
        while i < n:
            if s[i].isdigit():
                curr = ''
                while i < n and s[i].isdigit():
                    curr += s[i]
                    i += 1
                decodeStack.append(curr)
            elif s[i].isalpha():
                curr = ''
                while i < n and s[i].isalpha():
                    curr += s[i]
                    i += 1
                if decodeStack and decodeStack[-1].isalpha():
                    decodeStack[-1] += curr
                else:
                    decodeStack.append(curr)
            elif s[i] == ']':
                curr = decodeStack.pop()
                decodeStack.pop()
                curr *= int(decodeStack.pop())
                if decodeStack and decodeStack[-1].isalpha():
                    decodeStack[-1] += curr
                else:
                    decodeStack.append(curr)
                i += 1
            else:
                decodeStack.append(s[i])
                i += 1
        return decodeStack[-1]

File: Stack/394. Decode String/Solution.java
import java.util.*;

class Solution {
    public String decodeString(String s) {
        // N : len(s)
        // Time Complexity; O(N)
        // Space Complexity; O(N)
        int n = s.length();
        int i = 0;
        Stack<String> stack = new Stack<>();
        while (i < n) {
            if (Character.isDigit(s.charAt(i))) {
                String curr = "";
                while (i < n && Character.isDigit(s.charAt(i))) curr += s.charAt(i++);
                stack.push(curr);
            } else if (Character.isLetter(s.charAt(i))) {
                String curr = "";
                while (i < n && Character.isLetter(s.charAt(i))) curr += s.charAt(i++);
                if (!stack.isEmpty() && Character.isLetter(stack.peek().charAt(0))) curr = stack.pop() + curr;
                stack.push(curr);
            } else if (s.charAt(i) == ']') {
                String curr = stack.pop();
                stack.pop();
                curr = curr.repeat(Integer.parseInt(stack.pop()));
                if (!stack.isEmpty() && Character.isLetter(stack.peek().charAt(0))) curr = stack.pop() + curr;
                stack.push(curr);
                i++;
            } else stack.push(String.valueOf(s.charAt(i++)));
        }
        return stack.peek();
    }
}